File: src/main/java/com/projectswg/forwarder/services/server/ServerConnectionService.java
package com.projectswg.forwarder.services.server;

import com.projectswg.connection.HolocoreSocket;
import com.projectswg.connection.RawPacket;
import com.projectswg.forwarder.Forwarder.ForwarderData;
import com.projectswg.forwarder.intents.client.ClientDisconnectedIntent;
import com.projectswg.forwarder.intents.client.DataPacketInboundIntent;
import com.projectswg.forwarder.intents.client.DataPacketOutboundIntent;
import com.projectswg.forwarder.intents.control.StartForwarderIntent;
import com.projectswg.forwarder.intents.control.StopForwarderIntent;
import com.projectswg.forwarder.intents.server.RequestServerConnectionIntent;
import com.projectswg.forwarder.intents.server.ServerConnectedIntent;
import com.projectswg.forwarder.intents.server.ServerDisconnectedIntent;
import com.projectswg.forwarder.resources.networking.NetInterceptor;
import me.joshlarson.jlcommon.concurrency.BasicThread;
import me.joshlarson.jlcommon.control.IntentChain;
import me.joshlarson.jlcommon.control.IntentHandler;
import me.joshlarson.jlcommon.control.Service;
import me.joshlarson.jlcommon.log.Log;

public class ServerConnectionService extends Service {
	
	private static final int CONNECT_TIMEOUT = 5000;
	
	private final IntentChain intentChain;
	private final BasicThread thread;
	
	private HolocoreSocket holocore;
	private NetInterceptor interceptor;
	private ForwarderData data;
	
	public ServerConnectionService() {
		this.intentChain = new IntentChain();
		this.thread = new BasicThread("server-connection", this::primaryConnectionLoop);
		this.holocore = null;
		this.interceptor = null;
		this.data = null;
	}
	
	@Override
	public boolean stop() {
		return stopRunningLoop();
	}
	
	@IntentHandler
	private void handleStartForwarderIntent(StartForwarderIntent sfi) {
		interceptor = new NetInterceptor(sfi.getData());
		data = sfi.getData();
	}
	
	@IntentHandler
	private void handleStopForwarderIntent(StopForwarderIntent sfi) {
		stopRunningLoop();
	}
	
	@IntentHandler
	private void handleRequestServerConnectionIntent(RequestServerConnectionIntent rsci) {
		if (thread.isExecuting())
			stopRunningLoop();
		thread.start();
	}
	
	@IntentHandler
	private void handleClientDisconnectedIntent(ClientDisconnectedIntent cdi) {
		stopRunningLoop();
	}
	
	@IntentHandler
	private void handleDataPacketInboundIntent(DataPacketInboundIntent dpii) {
		HolocoreSocket holocore = this.holocore;
		if (holocore != null)
			holocore.send(interceptor.interceptClient(dpii.getData()));
	}
	
	private boolean stopRunningLoop() {
		thread.stop(true);
		return thread.awaitTermination(1000);
	}
	
	private void primaryConnectionLoop() {
		try (HolocoreSocket holocore = new HolocoreSocket(data.getAddress().getAddress(), data.getAddress().getPort())) {
			this.holocore = holocore;
			Log.t("Attempting to connect to server at %s", holocore.getRemoteAddress());
			if (!holocore.connect(CONNECT_TIMEOUT)) { 
				Log.w("Failed to connect to server!");
				return;
			}
			Log.i("Successfully connected to server at %s", holocore.getRemoteAddress());
			
			intentChain.broadcastAfter(getIntentManager(), new ServerConnectedIntent());
			while (holocore.isConnected()) {
				RawPacket inbound = holocore.receive();
				if (inbound == null) {
					Log.w("Server closed connection!");
					return;
				}
				intentChain.broadcastAfter(getIntentManager(), new DataPacketOutboundIntent(interceptor.interceptServer(inbound.getData())));
			}
		} catch (Throwable t) {
			Log.w("Caught unknown exception in server connection! %s: %s", t.getClass().getName(), t.getMessage());
			Log.w(t);
		} finally {
			Log.i("Disconnected from server.");
			intentChain.broadcastAfter(getIntentManager(), new ServerDisconnectedIntent());
			this.holocore = null;
		}
	}
	
}

File: src/main/kotlin/com/projectswg/forwarder/services/server/ServerConnectionService.kt
package com.projectswg.forwarder.services.server

import com.projectswg.common.network.packets.swg.holo.HoloConnectionStopped
import com.projectswg.forwarder.Forwarder.ForwarderData
import com.projectswg.forwarder.intents.*
import com.projectswg.forwarder.resources.networking.NetInterceptor
import com.projectswg.forwarder.resources.server.HolocoreConnection
import me.joshlarson.jlcommon.concurrency.BasicThread
import me.joshlarson.jlcommon.concurrency.ScheduledThreadPool
import me.joshlarson.jlcommon.control.IntentHandler
import me.joshlarson.jlcommon.control.Service
import me.joshlarson.jlcommon.log.Log
import java.util.concurrent.atomic.AtomicReference

class ServerConnectionService : Service() {
	
	private val thread = BasicThread("server-connection") { this.primaryConnectionLoop() }
	private val keepAliveThread = ScheduledThreadPool(1, "server-keepalive")
	
	private val currentConnection = AtomicReference<HolocoreConnection?>(null)
	
	private lateinit var interceptor: NetInterceptor // set by StartForwarderIntent
	private lateinit var data: ForwarderData         // set by StartForwarderIntent
	
	override fun start(): Boolean {
		keepAliveThread.start()
		keepAliveThread.executeWithFixedDelay(30_000, 30_000) { currentConnection.get()?.ping() }
		return super.start()
	}
	
	override fun stop(): Boolean {
		keepAliveThread.stop()
		val stoppedKeepAlive = keepAliveThread.awaitTermination(1000);
		val stoppedServerConnection = stopRunningLoop(HoloConnectionStopped.ConnectionStoppedReason.APPLICATION)
		
		return stoppedKeepAlive && stoppedServerConnection
	}
	
	@IntentHandler
	private fun handleStartForwarderIntent(sfi: StartForwarderIntent) {
		interceptor = NetInterceptor(sfi.data)
		data = sfi.data
	}
	
	@IntentHandler
	private fun handleStopForwarderIntent(sfi: StopForwarderIntent) {
		stopRunningLoop(if (data.crashed) HoloConnectionStopped.ConnectionStoppedReason.CRASH else HoloConnectionStopped.ConnectionStoppedReason.APPLICATION)
	}
	
	@IntentHandler
	private fun handleRequestServerConnectionIntent(rsci: RequestServerConnectionIntent) {
		if (currentConnection.get()?.getConnectionStatus() == HolocoreConnection.ServerConnectionStatus.CONNECTING)
			return  // It's trying to connect - give it a little more time
		
		if (stopRunningLoop(HoloConnectionStopped.ConnectionStoppedReason.NEW_CONNECTION))
			thread.start()
	}
	
	@IntentHandler
	private fun handleClientDisconnectedIntent(cdi: ClientDisconnectedIntent) {
		stopRunningLoop(HoloConnectionStopped.ConnectionStoppedReason.APPLICATION)
	}
	
	@IntentHandler
	private fun handleDataPacketInboundIntent(dpii: DataPacketInboundIntent) {
		val data = dpii.data
		if (data.size < 6) {
			return // not a valid packet
		}
		
		currentConnection.get()?.sendPacket(data)
	}
	
	private fun stopRunningLoop(reason: HoloConnectionStopped.ConnectionStoppedReason): Boolean {
		if (!thread.isExecuting)
			return true
		
		currentConnection.get()?.setDisconnectReason(reason)
		Log.d("Terminating connection with the server. Reason: $reason")
		thread.stop(true)
		return thread.awaitTermination(5000)
	}
	
	private fun primaryConnectionLoop() {
		try {
			HolocoreConnection(intentManager, interceptor, data).use {
				currentConnection.set(it)
				it.handle()
				currentConnection.set(null)
			}
		} catch (t: Throwable) {
			Log.w("Caught unknown exception in server connection! %s: %s", t.javaClass.name, t.message)
			Log.w(t)
		}
	}
	
}
